api/turing/batch/ensembling/runner.go

```diff
@@ -96,6 +100,18 @@ func (r *ensemblingJobRunner) updateStatus() {
 		RetryCountLessThan: &r.maxRetryCount,
 	}
 	r.processEnsemblingJobs(optionsJobBuildingImage)
+
+	// JobTerminating
+	optionsJobTerminatingImage := service.EnsemblingJobListOptions{
```

Comment:
Just curious: do we need to maintain some sort of queue to process jobs? I.e. job that was processed (but somehow would still need to be processed again during the next run) should be appended to the end of queue and runner processes only `batchSize` jobs from the top of queue? Also, if there are more than `batchSize` jobs waiting to be processed, shouldn't we just try to process them all before runner falls asleep before the next invocation?

Reply:
I did consider building an in memory queue (I suppose this is what you mean; Merlin does this) but I think it's too much effort for not much gain. I decided on an external queue only (postgres) during the batch orchestration PR. I would argue on the following points:

1. A 10 second (configurable, we can even put it as short as 2-3 seconds) delay is not going to make a difference.
2. We still have to maintain an external state because of unforeseen downtimes.

> Also, if there are more than batchSize jobs waiting to be processed, shouldn't we just try to process them all before runner falls asleep before the next invocation?

Seems like I forgot to rename `batch_size` to `RecordsToProcessInOneIteration` but I did in the config.

Actually, there can be, at one instance, more than `RecordsToProcessInOneIteration` processing at once. If you notice [here](https://github.com/ashwinath/turing/blob/terminate-ensembling-job/api/turing/batch/runner/base.go#L18), it doesn't actually wait for the job to be complete before the next cycle starts. I made sure of this because the consideration I had was not to overload postgres and kube-apiserver. Consider this example where  `RecordsToProcessInOneIteration` = 2:

```
--------------------------------------------------------------> time
---> fire job
    -------------------------------> job 1
    ----------------------------> job 2
<----------> ---> fire job
  Delay
                 -------------------------------> job 3
                 ----------------------------> job 4
                 <-------------->
            overlapping of 4 jobs here.
```
Which is to say that at most the delay will be based on `BatchRunnerConfig.TimeInterval`.

Reply:
> I decided on an external queue only (postgres)

How exactly does it work? Do you order records by `updated_at` and process first these jobs, that wait in `queue` the longest?

Reply:
There is no order by (although its probably good practice to put limit and offsets with order by). In any case this is inconsequential.

The query that picks up the records for processing is this `SELECT * FROM "ensembling_jobs"  WHERE (status IN ('pending','failed_submission','failed_building')) AND (retry_count < 3) LIMIT 10 OFFSET 0`, which is written ]here]( https://github.com/gojek/turing/blob/main/api/turing/batch/ensembling/runner.go#L196-L208.)

At different points of processing the ensembling job, the status changes, which the lifecycle of the finite state machine is stated [here](https://github.com/gojek/turing/blob/main/api/turing/models/ensembling_job.go#L86-L106). This is to ensure we do not double process a record.
```go
// Status is the state of the finite machine ensembling job.
// Possible statuses:
// JobPending --▶ JobFailedSubmission
//     |
//     |
//     |
//     |
// JobBuildingImage --▶ JobFailedBuildImage
//     |
//     |
//     |
//     |
//     ▼
// JobRunning --▶ JobFailed
//     |
//     |
//     |--▶ JobTerminated
//     |
//     |
//     ▼
// JobCompleted
type Status string
```

There is no queue, it's constantly being processed or waiting for I/O. The process can be broken down as follows in pseudocode:

```
while true:
  fire and forget runner <-------- point A
  wait
  
runner():
  processJob
  updateStatus
  
processJob():
  query records
  build image
  wait for build image to be done (blocking operation)
  create secrets/svc acc/rbac etc
  create spark job (don't wait, same as kubectl apply -f sparkapplication.yaml)
  
updateStatus():
  query based on conditions() // be it terminating, imagebuilding or running
  check the status on Kubernetes
  update postgres record  
```

I'm not sure what you mean by waiting. The only waiting that is done is between firing the runners. which is explained at point A above. The actual code is [here](https://github.com/ashwinath/turing/blob/terminate-ensembling-job/api/turing/batch/runner/base.go#L15-L18)

```go
// RunBatchRunners will run all runners asynchronously.
func RunBatchRunners(interval time.Duration, runners []BatchJobRunner) {
	for {
		for _, runner := range runners {
			go runner.Run()
		}
		time.Sleep(interval)
	}
}
```

Since `go runner.Run()` is not a blocking operation, it does a `time.Sleep(interval)` almost immediately. Which allows for the regular firing of batch runners.

The "queue" that we speak of is just the database, we just keep polling the database to check for `pending`/`failed_submission`/`failed_building` records for processing.

Reply:
The only difference I can see for the order by is either employing a stack/queue or some other picking strategy which I don't think is needed yet or any foreseeable future.

Reply:
Ok, thanks for the detailed explanation. I'm not too concerned about it for now.